### PR TITLE
Client-side entitlement caching

### PR DESCRIPTION
**Pull request by clemire**
Added a wrapped TTL cache (with unit tests) and using it in the SpacDapp to cache entitlements and entitlement evaluations. For local dev, the positive ttl is shortened to 5s for ease of testing.

### CODE REVIEW

File: packages/web3/src/EntitlementCache.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+            return { value: negativeCacheResult, cacheHit: true }
+        }
+
+        const positiveCacheResult = this.positiveCache.get(key)
+        if (positiveCacheResult !== undefined) {
+            return { value: positiveCacheResult, cacheHit: true }
+        }
+
+        const value = await onCacheMiss(keyable)
+        if (!!value == false) {
```
Comment (brianathere) [inline packages/web3/src/EntitlementCache.ts:53]: Won't this collapse into the same ring when you want it to be an error or a minimum number?

**Review by brianathere — APPROVED** [review]

File: packages/web3/src/EntitlementCache.ts (hunk 1 above)
Comment (clemire) [reply packages/web3/src/EntitlementCache.ts:53]: I suppose yes, I think I misunderstood what the go code was doing and thought this was behaving the same when it was not.

**Review by salzbrenner — APPROVED** [review]

File: packages/web3/src/EntitlementCache.ts (hunk 1 above)
Comment (sergekh2) [inline packages/web3/src/EntitlementCache.ts:53]: Lets not use this pattern:
in JS this leads to all manner of side effects, when values evaluate to false. 
Instead, please change onCacheMiss to return either CacheResult or similar type and check boolean explicitly.

File: packages/web3/src/EntitlementCache.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
+import TTLCache from '@isaacs/ttlcache'
+
+export interface Keyable {
+    toKey(): string
+}
+
+export interface CacheResult<V> {
+    value: V
+    cacheHit: boolean
+    isPositive: () => boolean
```
Comment (sergekh2) [inline packages/web3/src/EntitlementCache.ts:10]: but why there is a need for the function, would a `boolean` be good enough?

**Review by sergekh2 — APPROVED** [review]

**Pull request merged**